modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj

```diff
@@ -755,41 +799,69 @@
-    {:group {stddev-name (aggregation->rvalue [:stddev expr])}
-     :post  {(annotate/aggregation-name ag) {:$pow [(str \$ stddev-name) 2]}}}))
+        stddev-name (name (gensym "$stddev-"))]
+    {:group {(subs stddev-name 1) (aggregation->rvalue [:stddev expr])}
+     :post  [{(annotate/aggregation-name ag) {:$pow [stddev-name 2]}}]}))
 
 (defmethod expand-aggregation :default
   [ag]
   {:group {(annotate/aggregation-name ag) (aggregation->rvalue ag)}})
 
+(def ^:private group-op
+  #{:var :share :count :avg :stddev :sum :min :max :distinct :sum-where :count-where})
+
+(defn- extract-aggregation [aggregation aggr-name]
+  (when (and (vector? aggregation) (seq aggregation))
+    (let [[op & args] aggregation]
+      (cond
+        (= op :aggregation-options)
+        (let [[embedding-expr aggregation'] (extract-aggregation (first args) aggr-name)]
+          [embedding-expr (into [:aggregation-options aggregation'] (rest args))])
+        (group-op op) [(str \$ aggr-name) aggregation]
+        :else (let [ges (map #(extract-aggregation % aggr-name) args)
+                    [embedding-expr aggregation'] (first (filter some? ges))]
+                (when-not aggregation'
+                  (throw
+                   (ex-info (tru "Don''t know how to handle aggregation {0}" aggregation)
+                            {:type :invalid-query, :clause aggregation})))
+                [(into [op] (map (fn [arg ge] (if ge embedding-expr arg)) args ges))
+                 aggregation'])))))
+
+(defn- expand-embedded-aggregation [aggregation]
+  (let [aggr-name (annotate/aggregation-name aggregation)
+        [embedding-expr aggregation-expr] (extract-aggregation aggregation aggr-name)
+        expanded (expand-aggregation aggregation-expr)]
+    (cond-> expanded
+      (not (string? embedding-expr))
+      (update :post conj {aggr-name (->rvalue embedding-expr)}))))
+
 (defn- group-and-post-aggregations
   "Mongo is picky about which top-level aggregations it allows with groups. Eg. even
    though [:/ [:count-if ...] [:count]] is a perfectly fine reduction, it's not allowed. Therefore
    more complex aggregations are split in two: the reductions are done in `$group` stage after which
    we do postprocessing in `$addFields` stage to arrive at the final result. The intermittent results
    accrued in `$group` stage are discarded in the final `$project` stage."
   [id aggregations]
-  (let [expanded-ags (map expand-aggregation aggregations)
+  (let [expanded-ags (map expand-embedded-aggregation aggregations)
         group-ags    (mapcat :group expanded-ags)
         post-ags     (mapcat :post expanded-ags)]
-    [{$group (into (ordered-map/ordered-map "_id" id) group-ags)}
-     (when (not-empty post-ags)
-       {:$addFields (into (ordered-map/ordered-map) post-ags)})]))
+    (into [{$group (into (ordered-map/ordered-map "_id" id) group-ags)}]
+          (map (fn [p] {:$addFields p}))
+          post-ags)))
```

Comment:
Create separate groups so that values introduced by previous stages can be referred to.